util/network-devp2p/src/node_table.rs

```diff
@@ -235,47 +236,100 @@ impl Hash for Node {
 	}
 }
 
-const MAX_NODES: usize = 1024;
+pub const MAX_NODES_IN_TABLE: usize = 4096;
+const MAX_NODES_IN_FILE: usize = 1024;
 const NODES_FILE: &str = "nodes.json";
 
 /// Node table backed by disk file.
 pub struct NodeTable {
 	nodes: HashMap<NodeId, Node>,
+	ordered_ids: Vec<NodeId>,
 	useless_nodes: HashSet<NodeId>,
 	path: Option<String>,
 }
 
 impl NodeTable {
 	pub fn new(path: Option<String>) -> NodeTable {
+		let nodes = NodeTable::load(path.clone());
+		let ordered_ids = NodeTable::make_ordered_entries(&nodes).iter().map(|m| m.id).collect();
 		NodeTable {
-			path: path.clone(),
-			nodes: NodeTable::load(path),
+			path,
+			nodes,
 			useless_nodes: HashSet::new(),
+			ordered_ids
 		}
 	}
 
 	/// Add a node to table
 	pub fn add_node(&mut self, mut node: Node) {
 		// preserve node last_contact
 		node.last_contact = self.nodes.get(&node.id).and_then(|n| n.last_contact);
-		self.nodes.insert(node.id, node);
+		let id = node.id;
+		if self.ordered_ids.len() == MAX_NODES_IN_TABLE {
+			self.nodes.remove(&self.ordered_ids.pop().expect("ordered_ids is not empty; qed"));
+		};
+		let index = self.get_index_to_insert(node.last_contact);
+		if self.nodes.insert(node.id, node).is_none() {
+			self.ordered_ids.insert(index, id);
+		};
 	}
 
-	/// Returns a list of ordered nodes according to their most recent contact
-	/// and filtering useless nodes. The algorithm for creating the sorted nodes
-	/// is:
+	/// Get index in the ordered entries vector to insert node based on its last contact value
+	fn get_index_to_insert(&self, last_contact: Option<NodeContact>) -> usize {
```

Comment:
Sorry, but I don't understand. Why are you introducing so complicated and fragile logic with keeping the order of nodes instead of just to limit the output number of ordered_entries?

Reply:
That's what I was thinking about in the beginning, and of course that would be the easiest solution. However I assume that having limited table, gives us more predictable behaviour from performance point of view. As I said after running node for 4 hrs the table had 10k records and was growing.

Reply:
Ok. Can you at least create a couple tests for this method? (I mean get_index_to_insert)